sbor/src/decode.rs

```diff
@@ -40,14 +29,17 @@ pub enum DecodeError {
 
 /// A data structure that can be decoded from a byte array using SBOR.
 pub trait Decode<X: CustomTypeId>: Sized {
+    /// Decodes from the byte array encapsulated by the given decoder.
     fn decode(decoder: &mut Decoder<X>) -> Result<Self, DecodeError> {
-        Self::check_type_id(decoder)?;
-        Self::decode_value(decoder)
+        let type_id = decoder.read_type_id()?;
+        Self::decode_with_type_id(decoder, type_id)
     }
 
-    fn check_type_id(decoder: &mut Decoder<X>) -> Result<(), DecodeError>;
-
-    fn decode_value(decoder: &mut Decoder<X>) -> Result<Self, DecodeError>;
+    /// Decodes from the byte array encapsulated by the given decoder, with a preloaded type id.
+    fn decode_with_type_id(
```

Comment:
For symmetry, should this be called `decode_body_with_type_id`?

Reply:
Yes!